readthedocs/search/parsers.py

```diff
@@ -21,6 +21,68 @@ def __init__(self, version):
         self.project = self.version.project
         self.storage = get_storage_class(settings.RTD_BUILD_MEDIA_STORAGE)()
 
+    def _get_page_content(self, page):
+        """Gets the page content from storage."""
+        content = None
+        try:
+            storage_path = self.project.get_storage_path(
+                type_='html',
+                version_slug=self.version.slug,
+                include_file=False,
+            )
+            file_path = self.storage.join(storage_path, page)
+            with self.storage.open(file_path, mode='r') as f:
+                content = f.read()
+        except Exception:
+            log.warning(
+                'Unhandled exception during search processing file: %s',
+                page,
+            )
+        return content
+
+    def _get_page_title(self, body, html):
+        """
+        Gets the title from the html page.
+
+        The title is the first section in the document,
+        falling back to the ``title`` tag.
+        """
+        first_header = body.css_first('h1')
+        if first_header:
+            title, _ = self._parse_section_title(first_header)
+            return title
+
+        title = html.css_first('title')
+        if title:
+            return self._parse_content(title.text())
+
+        return None
+
+    def _get_main_node(self, html):
```

Comment:
This is definitely part of the complexity that an "only HTML" parser runs into. We're definitely going to end up parsing random navigation and other content on pages, just because we're relying on heuristics instead of the tool telling us what the exact main content is. 

I don't think we can avoid that, but definitely a downside of generic HTML parsing.

Reply:
yeah, that's why I want just to write a guide for authors of theme/static site generators. That would also make their site to be better indexed by search engines and work better with screen readers.